File: leetcode/spiral_matrix/spiral_matrix.cpp
/*
 * Time Complexity: O(N * M)
 *   - N is the number of rows, M is the number of columns.
 *   - Each element is visited exactly once.
 *
 * Space Complexity: O(N * M)
 *   - For the output vector containing all elements.
 */

#include <vector>
#include <iostream>

using std::vector;

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int columnEnd = matrix.size(),
            rowEnd = matrix[0].size(),
            columnStart = 0,
            rowStart = 0,
            idx = 0;

        vector<int> ans(columnEnd * rowEnd);

        while(columnStart < columnEnd || rowStart < rowEnd) {
            for(int i = rowStart; i < rowEnd; i++) ans[idx++] = matrix[columnStart][i];
            columnStart++;

            if(idx >= ans.size()) break;

            for(int i = columnStart; i < columnEnd; i++) ans[idx++] = matrix[i][rowEnd-1];
            rowEnd--;

            if(idx >= ans.size()) break;

            for(int i = rowEnd-1; i >= rowStart; i--) ans[idx++] = matrix[columnEnd-1][i];
            columnEnd--;

            if(idx >= ans.size()) break;

            for(int i = columnEnd-1; i >= columnStart; i--) ans[idx++] = matrix[i][rowStart];
            rowStart++;

            if(idx >= ans.size()) break;
        }

        return ans;
    }
};

int main() {
    // Example usage
    Solution sol;
    vector<vector<int>> matrix = {{1, 2, 3, 4}, {5, 6, 7, 8}, {9, 10, 11, 12}};
    vector<int> result = sol.spiralOrder(matrix);
    
    for (int num : result) {
        std::cout << num << " ";
    }
    return 0;
}

File: leetcode/spiral_matrix/spiral_matrix.java
package spiral_matrix;

import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> ans = new ArrayList<>();
        int rowStart = 0,
            rowEnd = matrix.length,
            colStart = 0, 
            colEnd = matrix[0].length,
            allValues = rowEnd * colEnd;

        while(colStart < colEnd || rowStart < rowEnd) {
            for(int col = colStart; col < colEnd; col++) ans.add(matrix[rowStart][col]);
            rowStart++;

            if(ans.size() >= allValues) break;

            for(int row = rowStart; row < rowEnd; row++) ans.add(matrix[row][colEnd-1]);
            colEnd--;

            if(ans.size() >= allValues) break;

            for(int col = colEnd-1; col >= colStart; col--) ans.add(matrix[rowEnd-1][col]);
            rowEnd--;

            if(ans.size() >= allValues) break;

            for(int row = rowEnd-1; row >= rowStart; row--) ans.add(matrix[row][colStart]);
            colStart++;

            if(ans.size() >= allValues) break;
        }

        return ans;
    }
}
